# Compare Query Results

## Setup

In [8]:
import json
import pandas as pd
from IPython.display import HTML, display

queries_df = pd.read_csv("queries.csv")
queries_df = queries_df.set_index("query")

In [9]:
# Disclosure: Used Claude for crafting the table styling and formatting logic below.
def show_results(query, k=5):
    """Display pre-computed BM25 and semantic results from queries.csv side-by-side."""
    row = queries_df.loc[query]
    bm_results = json.loads(row["bm25"])[:k]
    sem_results = json.loads(row["semantic_search"])[:k]

    bm_rows = [
        f"**{r['title']}**<br><small>{r['author'] or '—'} · ⭐ {r['rating']:.1f} · score: {r['score']:.3f}</small>"
        for r in bm_results
    ]
    sem_rows = [
        f"**{r['title']}**<br><small>{r['author'] or '—'} · ⭐ {r['rating']:.1f} · score: {r['score']:.3f}</small>"
        for r in sem_results
    ]

    df = pd.DataFrame({
        "Rank": [f"#{i+1}" for i in range(k)],
        "BM25": bm_rows,
        "Semantic Search": sem_rows,
    }).set_index("Rank")

    display(HTML(df.to_html(escape=False)))

## BM25 vs. Semantic
Milestone 1: Qualitative Evaluation of Retrieval Methods, 4.3 Compare Results

### Query Set

| # | Query | Difficulty | Type |
|---|-------|-----------|------|
| 1 | `harry potter` | Easy | Keyword |
| 2 | `learn python programming beginner` | Easy | Keyword |
| 3 | `overcoming grief after losing a loved one` | Medium | Semantic |
| 4 | `self-help book for managing anxiety and stress at work` | Medium | Semantic |
| 5 | `historical novel about World War 2 from a civilian perspective` | Hard | Multi-attribute |

---

## Query 1 — Easy (Keyword): `harry potter`

In [10]:
show_results("harry potter")

,BM25,Semantic Search
Rank,,
#1,**Complete 1-st Edition Harry Potter Full Book Set Volumes 1-7 Hardback by J. K. Rowling**J.K. Rowling · ⭐ 4.6 · score: 20.378,**Fantastic Beasts and Where to Find Them**J.K. Rowling · ⭐ 4.7 · score: 0.992
#2,**Harry Potter y el legado maldito / Harry Potter and the Cursed Child (Spanish Edition)**J.K. Rowling · ⭐ 4.5 · score: 20.268,**Harry Potter: The Dark Arts (Tiny Book)**Insight Editions · ⭐ 4.5 · score: 1.039
#3,**كواليس هاري بوتر - Harry Potter Film Wizardry (Arabic Edition)**Brian Sibley · ⭐ 4.1 · score: 20.206,**Wizard's Hall**Jane Yolen · ⭐ 4.6 · score: 1.101
#4,**Harry Potter: Exploring Hogwarts ™ Sticky Note Tin Set (Set of 3)**— · ⭐ 4.4 · score: 19.466,**The Riddle of the New Testament. Faber & Faber. 1958.**— · ⭐ 5.0 · score: 1.116
#5,**Harry Potter: The Dark Arts (Tiny Book)**Insight Editions · ⭐ 4.5 · score: 18.555,**Complete 1-st Edition Harry Potter Full Book Set Volumes 1-7 Hardback by J. K. Rowling**J.K. Rowling · ⭐ 4.6 · score: 1.168


### Analysis

**BM25** correctly retrieves 5 books that specifically mention Harry Potter in the title.

**Semantic search** retrieves 2 books that specifically mention Harry Potter in the title, but also "Fanastic Beasts and Where to Find Them" which is part of the franchise but doesn't specifically mention Harry Potter in the title. "Wizard's Hall" is apparently a different book that came out before Harry Potter and is very similar, so we can see the semantic search finding similar titles here. The Riddle of the New Testament does not seem related though.

**Better: BM25** does better since it retrieves 5 related books, while semantic search only retrieves 3.

---

## Query 2 — Easy (Keyword): `learn python programming beginner`

In [11]:
show_results("learn python programming beginner")

,BM25,Semantic Search
Rank,,
#1,"**LEARN MICROPYTHON: Learn Basics Programming Hardware in MicroPython, ESP32 Programming, Arduino, Example Code, PWM, SPI Bus, I2C Bus and More**— · ⭐ 1.8 · score: 27.407",**101 Extra Python Challenges with Solutions / Code Listings**— · ⭐ 4.5 · score: 0.757
#2,**40 Algorithms Every Programmer Should Know: Hone your problem-solving skills by learning different algorithms and their implementation in Python**Imran Ahmad · ⭐ 4.3 · score: 24.897,"**LEARN MICROPYTHON: Learn Basics Programming Hardware in MicroPython, ESP32 Programming, Arduino, Example Code, PWM, SPI Bus, I2C Bus and More**— · ⭐ 1.8 · score: 0.808"
#3,**101 Extra Python Challenges with Solutions / Code Listings**— · ⭐ 4.5 · score: 23.131,**Hands-On Cryptography with Python: Leverage the power of Python to encrypt and decrypt data**— · ⭐ 3.6 · score: 1.084
#4,**Mastering Blender**Tony Mullen · ⭐ 4.3 · score: 19.483,**Sams Teach Yourself Data Structures and Algorithms in 24 Hours**Robert Lafore · ⭐ 4.8 · score: 1.115
#5,**Visual Basic 6**Harold Davis · ⭐ 3.9 · score: 19.314,"**Coding with Scratch 3.0: Workbook 3 (Sharp Series, Scratch)**Lex Sharp · ⭐ 4.2 · score: 1.127"


### Analysis

**BM25** finds 3 books that are specifically Python related, though only the first result seems beginner friendly. The last two results seem to be beginner programming related, but no Python related.

**Semantic search** leads with 101 Extra Python Challenges which is not beginner friendly but is Python related. After that is the same as the first result from BM25. The third result is python related but doesn't seem beginner friendly, and the fourth is beginner friendly but it's unclear whether it is specifically Python related. The fifth is related to coding but is for Scratch, not Python, and does not appear beginner friendly.

**Better: BM25** as it's top result is the closest match to the search query. Both algorithms struggled a lot with matching the "beginner" part of the query.

---

## Query 3 — Medium (Semantic): `overcoming grief after losing a loved one`

In [12]:
show_results("overcoming grief after losing a loved one")

,BM25,Semantic Search
Rank,,
#1,**Through It All**— · ⭐ 4.9 · score: 18.668,**Anatomy of Grief: An Inspirational Guide to Surviving the Death of Your Child**Barbara Repczynski · ⭐ 5.0 · score: 0.663
#2,**30 Days toward Healing Your Grief: A Workbook for Healing**— · ⭐ 4.4 · score: 18.046,**I'm Grieving as Fast as I Can: How Young Widows and Widowers Can Cope and Heal**Linda Sones Feinberg · ⭐ 4.5 · score: 0.849
#3,**Gravity's Embrace: A True Unsolved Mystery Surrounding An Alaskan Pilot**— · ⭐ 4.6 · score: 16.840,**Elf-Help Grief Therapy - Self-Help Encouragement 20178-ABBEY**— · ⭐ 4.5 · score: 0.907
#4,**I'm Grieving as Fast as I Can: How Young Widows and Widowers Can Cope and Heal**Linda Sones Feinberg · ⭐ 4.5 · score: 15.790,**You Will See Your Baby In Heaven: A Man's Perspective of Stillbirth**— · ⭐ 4.9 · score: 0.919
#5,**Naked and Not Ashamed**— · ⭐ 4.5 · score: 15.501,**The Birth We Call Death**Paul H. Dunn · ⭐ 4.7 · score: 0.925


### Analysis

**BM25** does return some grief related books (the 2nd and 4th are clearly grief related) but mostly the results seem unrelated. Interestingly, the 4th result appears the most related based on the title, yet it is only the 4th ranked. (Looking at semantic search, we can see that the same result is ranked 2nd.)

**Semantic search** gets all five results really spot on, returning several books very specifically about grieving after the loss of a loved one such as a child or partner.

**Better: Semantic search** clearly does a better job of encoding the meaning of the search phrase and comes up with very relevant results.

---

## Query 4 — Medium (Semantic): `self-help book for managing anxiety and stress at work`

In [13]:
show_results("self-help book for managing anxiety and stress at work")

,BM25,Semantic Search
Rank,,
#1,**The Art of Calm: Relaxation Through the Five Senses**Brian Luke Seaward · ⭐ 3.6 · score: 23.981,**The Art of Calm: Relaxation Through the Five Senses**Brian Luke Seaward · ⭐ 3.6 · score: 0.884
#2,**Emotional Core Therapy**Robert A Moylan · ⭐ 4.3 · score: 21.645,"**Cognitive Behavioral Therapy: A Psychologist's Guide to Overcome Anxiety, Depression, & Negative Thought Patterns: Psychology Self-Help, Book 5**— · ⭐ 3.8 · score: 0.927"
#3,**Loving Someone With Ocd: Help for You & Your Family**Cherry Pedrick · ⭐ 4.5 · score: 20.365,**Anxious for Nothing: Finding Calm in a Chaotic World**— · ⭐ 4.8 · score: 0.950
#4,**The Complete Credit Repair Kit + Cd-rom**Brette Sember · ⭐ 4.1 · score: 20.323,**2019 Daily Self-Care Diary: for BPD**— · ⭐ 3.0 · score: 0.986
#5,"**Mandalas and More Adult Coloring Book: Over 100 illustrations for stress relief, Autism, ADHD and depression. For adults and children. By Dr. Robert K. Wheeler Jr. (Fantastic Adult Coloring Books)**Dr. Robert K. Wheeler Jr. · ⭐ 5.0 · score: 19.064",**Emotional Core Therapy**Robert A Moylan · ⭐ 4.3 · score: 0.997


### Analysis

**BM25** top two recommendations seem very relevant to managing anxiety and stress based on their titles. I researched Emotional Core Therapy, the 3rd recommendation, which is about stress but appears to be more centered on relationships rather than work specifically. Loving Someone with OCD is not really a self-help book or about stress at work, and the complete credit repair kit is very off the mark. The last result is a coloring book for stress relief, which is related to the "stress" keyword but not much else in the query.

**Semantic search** picks the same top book, but it's recommendations for the 2nd and 3rd rank are much more relevant and clearly self-help style books. The 4th and 5th ranked books are less about stress and anxiety in the workplace, but are self-help type books about mental health in general. The 5th ranked book is the same as the 2nd ranked book in BM25.

**Better: Semantic search** does a better job, producing 3 good top matches while BM25 only produces one good top match.

---

## Query 5 — Hard (Multi-attribute): `historical novel about World War 2 from a civilian perspective`

In [14]:
show_results("historical novel about World War 2 from a civilian perspective")

,BM25,Semantic Search
Rank,,
#1,"**Caves, Cannons and Crinolines**Beverly Stowe McClure · ⭐ 4.3 · score: 20.037",**The World War Two Reader (Routledge Readers in History)**Gordon Martel · ⭐ 3.7 · score: 0.769
#2,"**Just War Reconsidered: Strategy, Ethics, and Theory (Battles and Campaigns Series)**— · ⭐ 4.5 · score: 18.401",**Modern Warfare**Ashley Brown · ⭐ 4.0 · score: 0.781
#3,**A High and Hidden Place: A Novel**Michele Claire Lucas · ⭐ 4.0 · score: 16.019,**BBC History Magazine:The Second World War Story**— · ⭐ 5.0 · score: 0.787
#4,"**Writings of a Rebel Colonel: The Civil War Diary and Letters of Samuel Walkup, 48th North Carolina Infantry**— · ⭐ 3.5 · score: 15.268",**The Meaning of the Second World War (Verso World History Series)**Ernest Mandel · ⭐ 4.2 · score: 0.796
#5,"**A Tale of Two Maidens: A Medieval French Story of Fate, Adventure, and the Hundred Years' War**Anne Echols · ⭐ 4.6 · score: 14.372",**Raj and Norah**— · ⭐ 4.6 · score: 0.810


### Analysis

**BM25** fails entirely at capturing the World War 2 piece of the query. Its rank 1 result, *Caves, Cannons and Crinolines*, is a Civil War novel, not about World War 2. It likely matched because of "historical," "novel," and "War" in its metadata. Rank 2, Just War Reconsidered, is an academic military ethics textbook. None of the five BM25 results are about world war 2, though they do mostly appear to be historical novels about war. It seems that BM25 captured "historical novel" and "war" from the search term but not World War 2 specifically. Interestingly, very few of the exact search terms appear in the titles of returned books (besides "war"), so it seems that BM25 is needing to go into the descriptions and reviews more to find relevant matches than in previous searches.

**Semantic search** correctly focuses on WWII across all five results but it seems to be pulling mostly academic books rather than novels. The 5th result, Raj and Norah, is a novel and is "a true story of love lost and found in WWII", which appears to actually be a very good match to the query, so it's odd that it is only in 5th place.

**Better: Neither** algorithm gets a very good match. Semantic search manages to capture World War 2 in all of it's results, but not historical novel. BM25 seems to capture the historical novel part of the query and that they are looking for a war theme, but does not return any results about World War 2 in particular. Neither method appears to have captured the "civilian perspective" piece of the search query.

## ___ LLM vs. ____ LLM
Final Submission, 1.2 Experiment with the LLM

### Query Set
| # | Query | Difficulty | Type |
|---|-------|-----------|------|
| 1 | `stephen king horror novel` | Easy | Keyword |
| 2 | `romance novel with enemies to lovers trope` | Medium | Keyword |
| 3 | `a story about a young person finding their identity` | Medium | Semantic |
| 4 | `mystery thriller with unreliable narrator set in small town` | Hard | Multi-attribute |
| 5 | `best science fiction book for someone who liked The Martian` | Hard | Multi-attribute |